src/test/protocol/TER_test.cpp

```diff
@@ -49,13 +52,214 @@ struct TER_test : public beast::unit_test::suite
         }
     }
 
+    // Helper template that makes sure two types are not convertible or
+    // assignable if not the same.
+    // o I1 one tuple index.
+    // o I2 other tuple index.
+    // o Tup is expected to be a tuple.
+    // It's a functor, rather than a function template, since a class template
+    // can be a template argument without being full specified.
+    template<std::size_t I1, std::size_t I2>
+    class NotConvertible
+    {
+    public:
+        template<typename Tup>
+        void operator()(Tup const& tup, beast::unit_test::suite&) const
+        {
+            // Entries in the tuple should not be convertible or assignable
+            // unless they are the same types.
+            using To_t = std::decay_t<decltype (std::get<I1>(tup))>;
+            using From_t = std::decay_t<decltype (std::get<I2>(tup))>;
+            static_assert (!std::is_same<From_t, To_t>::value ^
+                std::is_convertible<From_t, To_t>::value, "Convert err");
```

Comment:
`!a ^ b` seems like a strange way to write `a==b`. I feel like I'm missing something. Why the bit twiddling ops instead of `==`?

Reply:
Good catch.  Thanks.  The bit twiddling is because I was too deep into debugging the code to notice the equivalence.  `==` is much clearer.